### ISSUE TITLE
Handle unexpected 1XX status codes better.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8885]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8885|
|Type|defect|
|Created|2016-10-31 18:40:00Z|
|Branch|https://github.com/twisted/twisted/tree/8885-lukasa-103|

Long story short

Twisted's HTTP client does not tolerate non-100 or 101 status codes from the 1XX block in a sensible manner: it treats them as final responses, rather than as provisional ones.

Expected behaviour

When Twisted receives a 1XX status code that it does not recognise, it should either surface these to a user that expects them by means of a callback, or it should ignore them and only show the user the eventual final status code. This is required by RFC 7231 Section 6.2 (Informational 1xx), which reads:

> A client MUST be able to parse one or more 1xx responses received prior to a final response, even if the client does not expect one. A user agent MAY ignore unexpected 1xx responses.

Actual behaviour

Twisted treats the 1XX status code as final. It parses the header block as though it were a response in the 2XX or higher range. Twisted assumes that 1XX responses have no body, which means that it leaves the following 2XX response unconsumed in the socket buffer, which will ruin the framing for any subsequent HTTP transaction on the same socket, as well as hide critical data from the user.

Steps to reproduce

The following "server" can demonstrate the issue:

```
import socket
import time

document = b'''<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>title</title>
    <link rel="stylesheet" href="/other/styles.css">
    <script src="/other/action.js"></script>
  </head>
  <body>
    <h1>Hello, world!</h1>
  </body>
</html>
'''

s = socket.socket()
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind(('localhost', 8080))
s.listen(5)

while True:
    new_socket, _ = s.accept()
    data = b''

    while not data.endswith(b'\r\n\r\n'):
        data += new_socket.recv(8192)

    new_socket.sendall(
        b'HTTP/1.1 103 Early Hints\r\n'
        b'Server: socketserver/1.0.0\r\n'
        b'Link: </other/styles.css>; rel=preload; as=style\r\n'
        b'Link: </other/action.js>; rel=preload; as=script\r\n'
        b'\r\n'
    )
    time.sleep(1)
    new_socket.sendall(
        b'HTTP/1.1 200 OK\r\n'
        b'Server: socketserver/1.0.0\r\n'
        b'Content-Type: text/html\r\n'
        b'Content-Length: %s\r\n'
        b'Link: </other/styles.css>; rel=preload; as=style\r\n'
        b'Link: </other/action.js>; rel=preload; as=script\r\n'
        b'Connection: close\r\n'
        b'\r\n' % len(document)
    )
    new_socket.sendall(document)
    new_socket.close()
```

If this server is run directly, the following client can be used to test it:

```
from twisted.internet.task import react

import treq


def got_content(content):
    print content


def print_response(response):
    print(response.code, response.phrase)
    print(response.headers)

    return treq.text_content(response).addCallback(got_content)


def main(reactor, *args):
    d = treq.get('http://localhost:8080/')
    d.addCallback(print_response)
    return d


react(main, [])
```

This client will print

```
(103, 'Early Hints')
Headers({'content-encoding': [''], 'link': ['</other/styles.css>; rel=preload; as=style', '</other/action.js>; rel=preload; as=script'], 'server': ['socketserver/1.0.0']})

```

Note that the client has treated the 1XX response as final, and has left the 200 response unconsumed in the socket. That means that, thanks to Twisted's connection pooling, subsequent requests that re-use that TCP connection will have their framing ruined because of the unconsumed final response.

<details><summary>Searchable metadata</summary>

```
trac-id__8885 8885
type__defect defect
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__8885_lukasa_103 8885-lukasa-103
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1477939200258535 1477939200258535
changetime__1478603014322428 1478603014322428
version__None None
owner__Lukasa Lukasa

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_2"></a>|@Lukasa commented|
|-|-|

In changeset b05ed0f7d20462df701ebebaf93de3e6ba177e1d
```
#!CommitTicketReference repository="" revision="b05ed0f7d20462df701ebebaf93de3e6ba177e1d"
Failing tests for #8885.
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_3"></a>|@Lukasa commented|
|-|-|

In changeset 81e746b09e47729bd1cdc607104442d15f798a30
```
#!CommitTicketReference repository="" revision="81e746b09e47729bd1cdc607104442d15f798a30"
Add topfile for #8885.
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_4"></a>|@Lukasa commented|
|-|-|

A patch is available for review in https://github.com/twisted/twisted/pull/583

Some notes:

1. I suspect we should log out a message when we drop a 1XX status code like this, but I'm reluctant to contribute to the continued use of the old logging system in twisted.web. Thoughts?
2. I wondered whether we should continue to surface 101 responses as per their old logic. We might accidentally regress something if users are expecting to be delivered those responses. Any thoughts there?

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_5"></a>|@wsanchez commented|
|-|-|

reviewed

I suggest adding a log line with (old logging is fine, being in use already).  No need for re-review there.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @Lukasa|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_7"></a>|@Lukasa commented|
|-|-|

Thanks for the review wsanchez. Per your review I've gone back and added a log line and a test for that log, and assuming the tests all pass and coverage is up to scratch I'll go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_8"></a>|@Lukasa set status to `closed`|
|-|-|

In changeset 389b00a2b1b359c34c8fdd1877d9f9e214ea3979
```
#!CommitTicketReference repository="" revision="389b00a2b1b359c34c8fdd1877d9f9e214ea3979"
Merge 8885-lukasa-103: Ignore unexpected 1XX status codes.

Author: lukasa
Reviewer: wsanchez
Fixes: #8885
```
